### ISSUE TITLE
Generalize TieSpecifier.tie_consecutive_notes to accept a pattern

### ISSUE THREAD

**Issue by trevorbaca** [Abjad/abjad#601]
Right now `TieSpecifier.tie_consecutive_notes` accepts only a boolean: either all runs of consecutive notes are tied together or else no runs of consecutive notes are tied together. Here's an example:

```
tie_specifier = rhythmmakertools.TieSpecifier(
   tie_consecutive_notes=True,
   )
```

Extend `TieSpecifier.tie_consecutive_notes` to accept a `patterntools.Pattern` object:

```
tie_specifier = rhythmmakertools.TieSpecifier(
   tie_consecutive_notes=patterntools.select_every(indices=[0], period=2),
   )
```

The first site of extension will be `TieSpecifier.__init__()` to allow the new type of input. The second site of extension will be `TieSpecifier._do_tie_consecutive_notes()` to implement the new logic.

**Comment by jdavancens** [comment]
I have an idea for the first part which is to accept a `patterntools.Pattern` as default and in the case where `tie_consecutive_notes` is a boolean, cast it as a single element, inverted pattern with period=1.
Like this:
```
if tie_consecutive_notes is not None:
            if isinstance(tie_consecutive_notes, bool):
                tie_consecutive_notes = \
                    patterntools.select_all(tie_consecutive_notes)
            elif isinstance(tie_consecutive_notes, patterntools.Pattern):
                tie_consecutive_notes = tie_consecutive_notes
            else:
                message = 'tie_consecutive_notes must be boolean or pattern'
                raise Exception(message)
```

This alleviates having two different branches of logic in `TieSpecifier._do_tie_consecutive_notes`

For the second part, it's seems to make the most sense to split the groups of consecutive notes at the indices of the pattern, and apply tie spanners to those chunks whose length is greater than one, like:
```
if 1 < len(group):
    subgroups = some_splitting_function(group, self.tie_consecutive_notes)
    for subgroup in subgroups:
        if 1 < len(subgroup):
            attach(spannertools.Tie(), subgroup)

I'd have to invert the pattern to get the indices of the split points (basically where a logical tie would start). I'm not really sure the best, most Abjadian way to split a list into chunks at arbitrary indices. I think there's a NumPy function that does that, but adding that dependency seems against the Abjad philosophy.

**Comment by trevorbaca** [comment]
Joe, awesome!

Check sequencetools for fancy list manipulation!

**Comment by jdavancens** [comment]
I'm assuming the idea was to apply a tie pattern at the level of consecutive note groups.
Another possibility is apply the tie pattern to all the non-rest leaves in `divisions`

**Comment by jdavancens** [comment]
I was looking in sequencetools but nothing caught my eye.

**Comment by jdavancens** [comment]
nevermind found it!
partition_sequence_by_counts

**Comment by trevorbaca** [comment]
;-)

**Comment by jdavancens** [comment]
Any idea why pytest might not be working for me?

I'm running it from tools/rhythmmakertools.
I've updated to the most recent version of pytest.
Using Python 2.7 on a mac in a conda environment

It doesn't seem to be running any of the tests in each module in the folder
```
going into /Users/josephdavancens/git/abjad/abjad/tools/rhythmmakertools/test
========  test_rhythmmakertools_NoteRhythmMaker___call__.py  =========

=========  test_rhythmmakertools_NoteRhythmMaker___eq__.py  ==========

======  test_rhythmmakertools_TieSpecifier_split_at_indices.py  ======

=======  test_rhythmmakertools_TupletRhythmMaker___call__.py  ========

==========  test_rhythmmakertools_Pattern_matches_index.py  ==========

========  test_rhythmmakertools_DurationSpellingSpecifier.py  ========

=========  test_rhythmmakertools_TaleaRhythmMaker___eq__.py  =========

==  test_rhythmmakertools_TaleaRhythmMaker_read_talea_once_only.py  ==

========  test_rhythmmakertools_TaleaRhythmMaker___call__.py  ========

============  test_rhythmmakertools_BurnishSpecifier.py  =============

==  test_rhythmmakertools_BurnishSpecifier_outer_divisions_only.py  ==

==============  test_rhythmmakertools_TieSpecifier.py  ===============

=======  test_rhythmmakertools_IncisedRhythmMaker___call__.py  =======

====  test_rhythmmakertools_BeamSpecifier_beam_each_division.py  =====

====  test_rhythmmakertools_TaleaRhythmMaker_tie_split_notes.py  =====

no test dir found testing here: #####/abjad/abjad/tools/rhythmmakertools
*******************************************************************************
```

**Comment by jdavancens** [comment]
Ah py.test did the trick

**Comment by josiah-wolf-oberholtzer** [comment]
Don't forget to run `ajv doctest` from within `rhythmmakertools` too. A lot of our tests are doctests.

**Comment by jdavancens** [comment]
For sure

**Comment by jdavancens** [comment]
I think I got it!
I made some fairly extensive test functions that are all passing, and all doctests in `rhythmmakertools` are working.

**Comment by josiah-wolf-oberholtzer** [comment]
Awesome! Post a PR and mention this issue (GitHub should help you by autocompleting). Your PR link will show up in this thread.

One good test paradigm we've been using is to just compare the string format of some score component against a normalized string:

```
staff = Staff("c'4 d'4 e'4 f'4")
assert format(staff) == stringtools.normalize(
    '''
    \new Staff {
        c'4
        d'4
        e'4
        f'4
    }
    ''')
```

When you run `py.test` in verbose mode it'll give you an actual multiline diff of the two strings. Makes it really easy to find where things have changed.

**Comment by jdavancens** [comment]
trying to push my branch with `git push --set-upstream origin issue-601-generalize-tiespecifier`
but i get
```
remote: Permission to Abjad/abjad.git denied to jdavancens.
fatal: unable to access 'https://github.com/Abjad/abjad.git/': The requested URL returned error: 403
```
We tried `git push -f` last time, but that gives
`fatal: The current branch issue-601-generalize-tiespecifier has no upstream branch.`

**Comment by josiah-wolf-oberholtzer** [comment]
Did you fork Abjad, and are you working from that fork, or did you clone the official Abjad repo and are currently trying to push back to it?

tl;dr you don't have push privileges to Abjad, hence the permission denied.

What to do?

- Fork Abjad: github.com/jdavencens/abjad
- Fire up terminal and navigate via `cd` until you find your current Abjad repo.
- Point that repo at your fork:
  `git remote set-url origin git@github.com:jdavencens/abjad.git`
- Push your already made commits
- Visit `github.com/jdavencens/abjad`
- Make a PR!

**Comment by trevorbaca** [comment]
Still valid in June 2017.
